site-src/geps/gep-1426.md

```diff
@@ -125,26 +120,41 @@ An alternate pattern additionally supported by this approach would be to target
-    * The namespace colocation requirement effectively forecloses using this strategy to delegate route management, but there is service mesh [user interest](https://github.com/istio/istio/issues/22997) in merging separately-defined routes together to manage large configurations, multiple versions or canary deployments for a single mesh service.
-* NOT already specified as a `backendRef` of any _other_ `HTTPRoute` with a `Service` `parentRef`
-   * This restriction is intended to prevent route "inclusion/delegation" patterns or circular references from multiple tiers of routing rules _within mesh configuration_.
-   * A `Service` MAY still be listed as a `backendRef` on a different `HTTPRoute` which only targets a `Gateway` as a `parentRef`. The specific behavior of how Gateways might interact with mesh routing rules will be defined in a future GEP.
-    * A `Service` MAY still be listed as a `backendRef` on the _same_ `HTTPRoute` where it is specified as a `parentRef`, e.g. to only split off some subset of traffic from the default backend endpoints.
-    * This could be computationally expensive to validate.
+* Service producers, who want to modify behavior of *inbound* requests to their services
+* Service consumers, who want to modify behavior of *outbound* requests to other services.
 
-This should be enforced by controller validation, which should set an `Accepted: { status: False }` condition with a corresponding new `ParentRefConflict` `RouteConditionReason`.
+While these concepts are not directly exposed in the API, a route is implicitly fulfilling one of these roles and behaves differently depending on the role.
 
-The logic described here is a bit verbose, but could be validated with conformance tests, and is intended to allow an `HTTPRoute` configuring mesh traffic rules to additionally specify a `Gateway` `parentRef` (or `HTTPRoute` `parentRef` if [GEP-1058: Route Inclusion and Delegation](https://github.com/kubernetes-sigs/gateway-api/pull/1085) moves forward) to apply the same routing rules when exposing a mesh service outside the cluster.
+A route is a producer route when the `parentRef` refers to a service in the *same namespace*. This route SHOULD apply to all incoming requests to the service, including from clients in other namespaces.
+
+Note: Some implementations may only be able to apply routes on client-side proxies. As a result, these will likely only apply to requests from clients who are also in the mesh.
+
+A route is a consumer route when the `parentRef` refers to a service in *another namespace*. Unlike producer routes, consumer routes are scoped only the same namespace. This ensures that for traffic between two namespaces, another unrelated namespace cannot modify their traffic.
```

Comment:
Some thoughts on this from a "how Kuma does routes right now" perspective, sorry for the wall. Some of this probably belonged on the original GEP PR but it's still relevant to this PR especially:

One way to think about an `HTTPRoute` is by being explicit about its effects on a request based on the source and the destination of that request.

When an `HTTPRoute` attaches to a `Gateway` via `parentRef`, this attachment expresses both the source as well as the destination of a request. The source being "somewhere, I don't know this is the edge" and the destination being the listener or the listener's `hostname`. And there's the option of further restricting this in the HTTPRoute `hostname`.

`Gateway` also gives us `.allowedRoutes.namespaces` separately to answer the question "who can attach `HTTPRoutes` and affect requests that come to my listener?".

With this GEP as is, using a `Service` as `parentRef` only tells us something about the destination of the requests we want to alter.

In the mesh case, it's important to be able to talk about the _source_ of a request when doing traffic routing. That's what the idea of consumer/producer routes in this PR expresses. A consumer route is "only requests from this namespace" and a producer route is "requests from anywhere in the mesh". 

The PR gives the `Namespace` of an `HTTPRoute` the ability to scope a route based on source of a request. But doing this also answers the question of who can attach to a `Service` as a consumer/producer.

Is there a way to answer these two questions less implicitly/separately?

I introduce the following as an illustration of what I vaguely have in mind, not a concrete suggestion. The `Mesh` resource has already been discussed and shows up in some of the GEP alternatives. Some of the GEP alternatives may even address the following but I couldn't find this source & destination aspect mentioned explicitly in the GEP.

Let's just imagine a different `parentRefs` and specify exactly two object references. The first to scope the route based on the request's _source_ and the other on the request's _destination_. Theoretically this is compatible with `parentRefs` as is, using a tuple as a list. I'll illustrate with a producer and a consumer route:

```
parentRefs:
  from:
    kind: Mesh
  to:
    kind: Service
    name: my-service
    namespace: my-namespace
```

This is a producer route which would affect requests _from_ anywhere in the mesh _to_ the `my-service` `Service`.

Translated to this PR IIUC:

```
metadata:
  namespace: my-service
parentRefs:
  - kind: Service
    name: my-service
    namespace: my-namespace
```

And then a consumer route:

```
parentRefs:
  from:
    kind: Namespace
    name: other-namespace
  to:
    kind: Service
    name: my-service
    namespace: my-namespace
```

that would affect requests _from_ services in `other-namespace` _to_ the `my-service` `Service`.

Translated to this PR:

```
metadata:
  namespace: other-namespace
parentRefs:
  - kind: Service
    name: my-service
    namespace: my-namespace
```

Coming from Kuma at least, the "object refs as source and destination" style feels more natural when thinking about which requests are affected by a route. "Which namespace is my route in plus an object ref" feels a little implicit/subtle.

If we were designing mesh routing as a Policy Attachment policy, we might have come up with something like

```
spec:
  targetRef:
    kind: Mesh
  to:
    kind: Service
  ...
```

Explicit refs would also give the option to specify something like:

```
parentRefs:
  from:
    kind: Service
    name: other-service
    namespace: other-namespace
  to:
    kind: Service
    name: my-service
    namespace: my-namespace
```

to scope a route to apply to requests from and to single services.

If the producer route only ever needs to be created in the same `Namespace` as the `Service`, a consumer route _never_ needs to be created in the same `Namespace` as the `Service` and `Namespace` is the only answer to "who's allowed to be a producer and who a consumer", then this degenerates to what the PR proposes.

It's different and I don't mean to open old discussion but it could be a useful perspective on how to deal with the consumer vs producer dichotomy.

Reply:
I do generally agree with ~everything you said. Some initial reactions:

* IMO producer routes ought to always be in the same namespace as the service. Otherwise they are not really the "producer"?
* Wanting "same namespace consumer" or "more specific than namespace" is definitely valid though
* `From=specific Service` is suspicious, since requests come from workloads not Services. But can be made to work or slightly tweaked of course

In general I would love to have a reasonable way to be more specific, I am just not sure the best schema to represent that

Reply:
* Yeah I think I mostly agree about producer routes. Though if we consider that producer routes come from the producer role, it makes me wonder if there's a use case for GEP-1324 mesh administrator routes, which might reside in a different "administrator namespace". The only cases I can think of are especially, but also only, useful with `backendRefs` omitted. And probably also need more generality in the `to` sense than a `Service` `parentRef`.

  Looking at GEP-1426 I see under goals:

  > SHOULD allow control by the cluster operator (mesh administrator) to grant permission for whether HTTPRoute resources in a given namespace are allowed to configure mesh traffic routing.

  This doesn't feel covered yet, right?

* `Service` in `from` is a little suspicious, you're right. I guess it'd make sense only if a workload is pointed to by exactly one service backend `Service`? But yeah, like you said, `Service` in particular is probably not that important, it's about having some way to refer to "a workload" or "something more specific than namespace."

If tweaking `HTTPRoute` isn't in the cards, which is totally valid given it's primary use case not being mesh, I guess the too late for this PR, concrete suggestion would be a vote to reconsider/also support a new resource as `parentRef`. One that holds the `from` and `to` references. I see both `ServiceBinding` and `ServiceProjection` under GEP-1426 but I'm not totally sure either of those covers it.

Reply:
@michaelbeaumont thanks for the concise write-up! I wanted to comment here since the conversation in this thread has been slightly on-point with some of the concerns we've had in the Linkerd team.

On the whole, we're pretty much in agreement with the GEP, we would like to discuss something similar to the `to` and `from` idea you wrote down though. On the Linkerd side, it will be necessary in the future to limit this configuration to only a subset of the workloads in the consumer namespace -- we already know that there are situations where users of Linkerd will need to enforce different policies for consumers in the same namespace.

We were thinking this could be addressed for a resource that captures the Consumer-to-Service binding relationship. Something between `ServiceBinding` and `ServiceProjection` (although projection probably fits better as a model and name). For example:

```
spec:
  consumerRefs:
    - group: apps
      kind: Deployment
      name: web
  providerRef:
    kind: Service
    namespace: backend
    name: emoji
status:
  conditions: ...
```

The resource describes traffic originating from a consumer workload (`deploy/web`) to a producer service (`service/emoji`). The status is added in for conflict resolution. Wdyt about an approach similar to this?

cc: @kflynn

Reply:
I think this is a great discussion and I don't want to bury it, so I opened an issue in https://github.com/kubernetes-sigs/gateway-api/issues/1640 to track this further. Based on the meeting it sounds like the desire is to move forward with this as-is to check something in as experimental, but we should follow up on exploring this further.